webui/react/src/components/NotebookModal.tsx

```diff
@@ -0,0 +1,309 @@
+import { Button, InputNumber, Modal } from 'antd';
+import { Form, Input, Select } from 'antd';
+import { ModalProps } from 'antd/es/modal/Modal';
+import yaml from 'js-yaml';
+import React, { Dispatch, useCallback, useEffect, useReducer, useState } from 'react';
+
+import useStorage from 'hooks/useStorage';
+import { getResourcePools, getTaskTemplates } from 'services/api';
+import { NotebookConfig, RawJson, ResourcePool, ResourceType, Template } from 'types';
+import { launchNotebook, previewNotebook } from 'utils/task';
+
+import Link from './Link';
+import css from './NotebookModal.module.scss';
+import RadioGroup from './RadioGroup';
+import Spinner from './Spinner';
+
+const MonacoEditor = React.lazy(() => import('react-monaco-editor'));
+
+const { Option } = Select;
+const { Item } = Form;
+
+const STORAGE_PATH = 'notebook-launch';
+const STORAGE_KEY = 'notebook-config';
+
+type DispatchFunction =
+  (Dispatch<{
+    key: keyof NotebookConfig,
+    value: string | number | undefined
+  }>)
+
+function reducer(
+  state: NotebookConfig,
+  action: {key: keyof NotebookConfig, value: string | number | undefined},
+): NotebookConfig {
+  return { ...state, [action.key]: action.value };
+}
```

Comment:
**suggestion:** we can refactor this later, but take a look at the pattern we use for `useReducer` in `ExperimentVisualizationFilters.tsx` (particularly the `ActionType`, `Action` and `reducer`). It's a bit more verbose, but the biggest benefit of our existing pattern is that the dispatch is much more strongly typed and will prevent incorrect usages. For example, it's possible to set the `name` property of notebook config with a `number` type even though `name` is a `string` type. The other benefit is that the reducer can also perform more complex operations on the state change, though not needed in this case.